core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +320,99 @@ private void balance() {
             balanceByWeights();
         }
 
+        /**
+         * Makes a decision about moving a single shard to a different node to form a more
+         * optimally balanced cluster.  This method is invoked from the cluster allocation
+         * explain API only.
+         */
+        private RebalanceDecision balanceShard(final ShardRouting shard) {
+            if (shard.started() == false) {
+                // cannot rebalance a shard that isn't started
+                return RebalanceDecision.NOT_TAKEN;
+            }
+
+            Decision canRebalance = allocation.deciders().canRebalance(shard, allocation);
+            if (canRebalance.type() != Type.YES) {
+                return new RebalanceDecision(canRebalance, Type.NO, "rebalancing is not allowed");
+            }
+
+            if (allocation.hasPendingAsyncFetch()) {
+                return new RebalanceDecision(canRebalance, Type.NO, "cannot rebalance due to in-flight shard store fetches");
```

Comment:
maybe we should expand this text to explain why we don't allow balancing while in-flight shard store fetching is going on.